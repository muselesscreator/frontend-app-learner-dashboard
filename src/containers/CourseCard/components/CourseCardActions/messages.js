import { StrictDict } from 'utils';

export const messages = StrictDict({
  upgrade: {
    id: 'learner-dash.courseCard.actions.upgrade',
    description: 'Course card upgrade button text',
    defaultMessage: 'Upgrade',
  },
  beginCourse: {
    id: 'learner-dash.courseCard.actions.beginCourse',
    description: 'Course card begin-course button text',
    defaultMessage: 'Begin Course',
  },
  resume: {
    id: 'learner-dash.courseCard.actions.resume',
    description: 'Course card resume button text',
    defaultMessage: 'Resume',
  },
  viewCourse: {
    id: 'learner-dash.courseCard.actions.viewCourse',
    description: 'Course card view-course button text',
    defaultMessage: 'View Course',
  },
});

export default messages;
