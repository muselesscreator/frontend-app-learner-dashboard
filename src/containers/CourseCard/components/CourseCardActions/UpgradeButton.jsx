import React from 'react';
import PropTypes from 'prop-types';

import { Button } from '@edx/paragon';
import { Locked } from '@edx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import { hooks } from 'data/redux';
import messages from './messages';

export const UpgradeButton = ({ cardId }) => {
  const { upgradeUrl } = hooks.useCardCourseRunData(cardId);
  const { canUpgrade } = hooks.useCardEnrollmentData(cardId);
  const { isMasquerading } = hooks.useMasqueradeData();
  const { formatMessage } = useIntl();
  if (!upgradeUrl) {
    return null;
  }
  return (
    <Button
      iconBefore={Locked}
      variant="outline-primary"
      disabled={isMasquerading || !canUpgrade}
      as="a"
      href={upgradeUrl}
    >
      {formatMessage(messages.upgrade)}
    </Button>
  );
};
UpgradeButton.propTypes = {
  cardId: PropTypes.string.isRequired,
};
export default UpgradeButton;
