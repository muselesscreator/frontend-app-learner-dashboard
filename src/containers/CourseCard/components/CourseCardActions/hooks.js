import { Locked } from '@edx/paragon/icons';

import { selectors } from 'data/redux';
import { useIntl, getCardValues } from 'hooks';
import messages from './messages';

const { cardData } = selectors;

export const actionHooks = ({ courseNumber }) => {
  const { formatMessage } = useIntl();
  const data = getCardValues(courseNumber, {
    canUpgrade: cardData.canUpgrade,
    isAudit: cardData.isAudit,
    isAuditAccessExpired: cardData.isAuditAccessExpired,
    isVerified: cardData.isVerified,
    isPending: cardData.isCourseRunPending,
    isFinished: cardData.isCourseRunFinished,
  });
  let primary;
  let secondary = null;
  if (!data.isVerified) {
    secondary = {
      iconBefore: Locked,
      variant: 'outline-primary',
      disabled: !data.canUpgrade,
      children: formatMessage(messages.upgrade),
    };
  }
  if (data.isPending) {
    primary = { children: formatMessage(messages.beginCourse) };
  } else if (!data.isFinished) {
    primary = {
      children: formatMessage(messages.resume),
      disabled: data.isAudit && data.isAuditAccessExpired,
    };
  } else {
    primary = { children: formatMessage(messages.viewCourse) };
  }
  return { primary, secondary };
};

export default actionHooks;
