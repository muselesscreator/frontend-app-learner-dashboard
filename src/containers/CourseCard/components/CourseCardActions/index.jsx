import React from 'react';
import PropTypes from 'prop-types';

import { Button } from '@edx/paragon';

import hooks from './hooks';

export const CourseCardActions = ({ courseNumber }) => {
  const { primary, secondary } = hooks({ courseNumber });
  return (
    <>
      {(secondary !== null) && (
        <Button {...secondary} />
      )}
      <Button {...primary} />
    </>
  );
};
CourseCardActions.propTypes = {
  courseNumber: PropTypes.string.isRequired,
};

export default CourseCardActions;
