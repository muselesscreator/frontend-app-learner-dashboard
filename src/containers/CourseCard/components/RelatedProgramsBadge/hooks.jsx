import React from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';

import { StrictDict } from 'utils';
import { selectors } from 'data/redux';
import { useCardValues } from 'hooks';

import messages from './messages';
import * as module from './hooks';

const { cardData } = selectors;

export const state = StrictDict({
  isOpen: (val) => React.useState(val), // eslint-disable-line
});

export const useRelatedProgramsBadgeData = ({ courseNumber }) => {
  const [isOpen, setIsOpen] = module.state.isOpen(false);
  const { formatMessage } = useIntl();
  const { numPrograms } = useCardValues(courseNumber, {
    numPrograms: cardData.numRelatedPrograms,
  });

  let programsMessage = '';
  if (numPrograms) {
    programsMessage = formatMessage(
      numPrograms === 1 ? messages.badgeLabelSingular : messages.badgeLabelPlural,
      { numPrograms },
    );
  }

  return {
    numPrograms,
    programsMessage,
    isOpen,
    openModal: React.useCallback(() => setIsOpen(true), [setIsOpen]),
    closeModal: React.useCallback(() => setIsOpen(false), [setIsOpen]),
  };
};

export default useRelatedProgramsBadgeData;
