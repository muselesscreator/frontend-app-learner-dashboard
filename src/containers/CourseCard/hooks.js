import { selectors } from 'data/redux';

import { useIntl, getCardValues } from 'hooks';

import * as module from './hooks';
import messages from './messages';

const { cardData } = selectors;

export const accessMessage = ({ courseNumber }) => {
  const { formatMessage, formatDate } = useIntl();

  const data = getCardValues(courseNumber, {
    accessExpirationDate: cardData.courseRunAccessExpirationDate,
    isAudit: cardData.isAudit,
    isFinished: cardData.isCourseRunFinished,
    isAuditAccessExpired: cardData.isAuditAccessExpired,
    endDate: cardData.courseRunEndDate,
  });
  if (data.isAudit) {
    return formatMessage(
      data.isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
      { accessExpirationDate: formatDate(data.accessExpirationDate) },
    );
  }
  return formatMessage(
    data.isFinished ? messages.courseEnded : messages.courseEnds,
    { endDate: formatDate(data.endDate) },
  );
};

export const cardHooks = ({ courseNumber }) => {
  const { formatMessage } = useIntl();
  const data = getCardValues(courseNumber, {
    title: cardData.courseTitle,
    bannerUrl: cardData.courseBannerUrl,
    providerName: cardData.providerName,
  });

  return {
    title: data.title,
    bannerUrl: data.bannerUrl,
    providerName: data.providerName || formatMessage(messages.unknownProviderName),
    accessMessage: module.accessMessage({ courseNumber }),
    formatMessage,
  };
};

export default cardHooks;
