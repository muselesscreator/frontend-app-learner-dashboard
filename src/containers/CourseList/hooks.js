import React from 'react';
import { useDispatch } from 'react-redux';

import { useCheckboxSetValues } from '@edx/paragon';

import { StrictDict } from 'utils';
import { actions, hooks as appHooks } from 'data/redux';
import { ListPageSize, SortKeys } from 'data/constants/app';
import { RequestKeys } from 'data/constants/requests';

import * as module from './hooks';

export const state = StrictDict({
  sortBy: (val) => React.useState(val), // eslint-disable-line
});

export const useCourseListData = () => {
  const dispatch = useDispatch();
  const pageNumber = appHooks.usePageNumber();
  const [filters, setFilters] = useCheckboxSetValues([]);
  const [sortBy, setSortBy] = module.state.sortBy(SortKeys.enrolled);

  const { numPages, visible } = appHooks.useCurrentCourseList({
    sortBy,
    isAscending: true,
    filters,
    pageSize: ListPageSize,
  });
  const handleRemoveFilter = (filter) => () => setFilters.remove(filter);
  const setPageNumber = (value) => dispatch(actions.app.setPageNumber(value));
  const initIsPending = appHooks.useIsPendingRequest(RequestKeys.initialize);

  return {
    pageNumber,
    numPages,
    setPageNumber,
    visibleList: visible,
    filterOptions: {
      sortBy,
      setSortBy,
      filters,
      setFilters,
      handleRemoveFilter,
    },
    showFilters: filters.length > 0,
    initIsPending,
  };
};

export default useCourseListData;
