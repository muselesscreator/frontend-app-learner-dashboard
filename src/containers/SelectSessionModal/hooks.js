import React from 'react';
import { useDispatch } from 'react-redux';

import { useIntl } from '@edx/frontend-platform/i18n';

import { StrictDict } from 'utils';

import { hooks as appHooks, thunkActions } from 'data/redux';
import * as module from './hooks';
import { LEAVE_OPTION } from './constants';
import messages from './messages';

export const state = StrictDict({
  selectedSession: (val) => React.useState(val), // eslint-disable-line
});

export const useSelectSessionModalData = () => {
  const dispatch = useDispatch();
  const selectedCardId = appHooks.useSelectSessionModalData().cardId;
  const {
    availableSessions,
    isFulfilled,
    uuid,
  } = appHooks.useCardEntitlementData(selectedCardId);
  const { title: courseTitle } = appHooks.useCardCourseData(selectedCardId);
  const { formatMessage } = useIntl();
  const [selectedSession, setSelectedSession] = module.state.selectedSession(null);

  let header;
  let hint;
  if (isFulfilled) {
    header = formatMessage(messages.changeOrLeaveHeader);
    hint = formatMessage(messages.changeOrLeaveHint);
  } else {
    header = formatMessage(messages.selectSessionHeader, {
      courseTitle,
    });
    hint = formatMessage(messages.selectSessionHint);
  }
  const updateCallback = appHooks.useUpdateSelectSessionModalCallback;

  const handleSelection = ({ target: { value } }) => setSelectedSession(value);
  const handleSubmit = () => {
    if (selectedSession === LEAVE_OPTION) {
      return dispatch(thunkActions.requests.leaveEntitlementSession({ uuid }));
    }
    return dispatch(thunkActions.requests.updateEntitlementEnrollment({ uuid, courseId: selectedSession }));
  };

  return {
    showModal: selectedCardId != null,
    closeSessionModal: updateCallback(dispatch, null),
    openSessionModal: (cardId) => updateCallback(dispatch, cardId),
    showLeaveOption: isFulfilled,
    availableSessions,
    hint,
    header,
    selectedSession,
    handleSelection,
    handleSubmit,
  };
};

export default useSelectSessionModalData;
