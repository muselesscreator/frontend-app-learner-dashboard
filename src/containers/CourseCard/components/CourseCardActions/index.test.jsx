import React from 'react';
import { shallow } from 'enzyme';

import hooks from './hooks';
import CourseCardActions from '.';

jest.mock('./hooks', () => ({
  __esModule: true,
  default: jest.fn(),
}));

const courseNumber = 'test-course-number';

describe('CourseCard Actions component', () => {
  it('loads primary and secondary button props from hook', () => {
    const mockHook = (args) => ({
      primary: {
        prop1: 'primary-prop1',
        prop2: 'primary-prop2',
        children: 'primary-children',
        courseNumber: args.courseNumber,
      },
      secondary: {
        prop1: 'primary-prop1',
        prop2: 'primary-prop2',
        children: 'primary-children',
        courseNumber: args.courseNumber,
      },
    });
    hooks.mockImplementationOnce(mockHook);
    expect(shallow(<CourseCardActions courseNumber={courseNumber} />)).toMatchSnapshot();
  });
  it('does not render secondary button if null is returned for secondary props', () => {
    const mockHook = (args) => ({
      primary: {
        prop1: 'primary-prop1',
        prop2: 'primary-prop2',
        children: 'primary-children',
        courseNumber: args.courseNumber,
      },
      secondary: null,
    });
    hooks.mockImplementationOnce(mockHook);
    expect(shallow(<CourseCardActions courseNumber={courseNumber} />)).toMatchSnapshot();
  });
});
