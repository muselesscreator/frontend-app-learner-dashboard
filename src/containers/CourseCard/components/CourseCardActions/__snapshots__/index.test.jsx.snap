// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCard Actions component does not render secondary button if null is returned for secondary props 1`] = `
<div
  data-test-id="CourseCardActions"
>
  <Button
    courseNumber="test-course-number"
    prop1="primary-prop1"
    prop2="primary-prop2"
  >
    primary-children
  </Button>
</div>
`;

exports[`CourseCard Actions component loads primary and secondary button props from hook 1`] = `
<div
  data-test-id="CourseCardActions"
>
  <Button
    courseNumber="test-course-number"
    prop1="primary-prop1"
    prop2="primary-prop2"
  >
    primary-children
  </Button>
  <Button
    courseNumber="test-course-number"
    prop1="primary-prop1"
    prop2="primary-prop2"
  >
    primary-children
  </Button>
</div>
`;
